Implement get_disjoint_mut (previously get_many_mut)

NiklasJonsson opened the pull request:
`std::collections::HashMap` is adding `get_many_mut` in https://github.com/rust-lang/rust/issues/97601 and I'm working on replacing some uses of that hash map with this one where some of the code uses `get_many_mut` (currently only a single location in rustc) so I would appreciate it if `indexmap::IndexMap` could provide the same interface.

This code uses a decent amount of unsafe compared to the rest of the file and I saw the comment in lib.rs on having almost all unsafe code in raw.rs but I couldn't figure out how to move parts of my implementation there as it is mostly dealing with initializing the array. I tried to use `array::map` to avoid all unsafe but couldn't quite get it to work with the lifetimes and the lambda.

If you want this change, I'll also write some docs.

cuviper commented:
I think ideally we would want two things for this, especially the first:

1. Stabilizing `HashMap::get_many_mut` so we can be sure to match that API as much as possible.
2. Stabilizing `<[T]>::get_many_mut` (still new in rust-lang/rust#83608) to deal with all the `unsafe` aspects. We might *also* want to wrap that into our own `get_many_index_mut`, just like we have `get_index` and `get_index_mut` for slice-like indexing.

> I tried to use `array::map` to avoid all unsafe but couldn't quite get it to work with the lifetimes and the lambda.

I suspect part of this is because you should be using a raw pointer to offset/index each of the elements, otherwise you have the main `&mut [T]` aliasing each of the `&mut T` you pull out while you're working.

NiklasJonsson commented:
> Stabilizing HashMap::get_many_mut so we can be sure to match that API as much as possible.

Yeah, this makes a lot of sense. Is it okay to just leave the PR open until then?

> Stabilizing <[T]>::get_many_mut

Haven't seen this before but it would indeed fit pretty well. It seems to have been open for quite a while without resolution though. Also, if they opt for a result-based return value, that'd not match the hashmap api (as it is now) so we'd have to do conversions in the implementation.

> I suspect part of this is because you should be using a raw pointer to offset/index each of the elements, otherwise you have the main &mut [T] aliasing each of the &mut T you pull out while you're working. And of course, using raw pointers will still need unsafe to form the final references.

Yeah, exactly, the lambda captures the whole array lifetime mutably and can't be invoked multiple times. I did not consider pointer arithmetic though, I'll experiment a bit and see how it looks. Probably similar to the `<[T]>::get_many_mut` examples 🙂

NiklasJonsson commented on src/map.rs:
```diff
@@ -477,6 +477,39 @@ where
+            // SAFETY: OK to discard mutable borrow lifetime as each index is unique
+            #[allow(unsafe_code)]
+            unsafe {
+                &mut *(&mut entries[i].value as *mut V)
```
By the way, I did run miri on the tests like:
```
# rustup +nightly component add miri
cargo clean
cargo +nightly miri test
```
but I know miri doesn't catch all cases of undefined behaviour.

NiklasJonsson replied:
Actually, maybe I understand. The variable `entries` is `&mut [Bucket<K, V>]` so it is a mutable reference the whole slice but the array `out` also holds mutable references to individual elements, which alias with `entries`. Is this what you are referring to?

cuviper replied:
Yeah, that's it. I'm a little surprised that miri doesn't see any problem, but I think it's because slice indexing is an intrinsic operation in MIR, so miri doesn't really see any use of the entire `entries`. If you force it to go through the `IndexMut` trait, `entries.index_mut(i)`, then it does complain:

```
test map::tests::many_mut_multi_success ... error: Undefined Behavior: trying to reborrow <390898> for SharedReadOnly permission at alloc159164[0xc], but that tag does not exist in the borrow stack for this location
    --> /home/jistone/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/cmp.rs:1531:27
     |
1531 |             PartialEq::eq(*self, *other)
     |                           ^^^^^
     |                           |
     |                           trying to reborrow <390898> for SharedReadOnly permission at alloc159164[0xc], but that tag does not exist in the borrow stack for this location
     |                           this error occurs as part of a reborrow at alloc159164[0xc..0x10]
     |
     = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
     = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <390898> was created by a retag at offsets [0xc..0x10]
    --> src/map.rs:502:19
     |
502  |           let out = indices.map(|i| {
     |  ___________________^
503  | |             // SAFETY: OK to discard mutable borrow lifetime as each index is unique
504  | |             #[allow(unsafe_code)]
505  | |             unsafe {
506  | |                 &mut *(&mut entries.index_mut(i).value as *mut V)
507  | |             }
508  | |         });
     | |__________^
help: <390898> was later invalidated at offsets [0x0..0x40]
    --> src/map.rs:506:29
     |
506  |                 &mut *(&mut entries.index_mut(i).value as *mut V)
     |                             ^^^^^^^^^^^^^^^^^^^^
     = note: backtrace: [...]
    --> src/map/tests.rs:449:5
     |
449  |     assert_eq!(map.get_many_mut([&1, &1123]), Some([&mut 10, &mut 100]));
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

> > We probably need a manual bounds-check too...
> 
> Yeah that makes sense if pointers are the way to go but it should still panic rather than return None, I assume?

Yeah, if `get_index_of` returns a bad index, then we have internal errors, so a panic is fine.

cuviper commented on src/map.rs:
```diff
@@ -477,6 +477,37 @@ where
+        for idx in indices {
+            let idx = idx?;
+            if idx >= len {
+                panic!("Index is out of range! Got '{idx}' but length is '{len}'")
```
Well, since this is internal "shouldn't happen" territory, I would also be fine with letting such bugs fall through to `get_many_index_mut -> None`. Maybe just add a `debug_assert!(idx < len)` for good measure.

NiklasJonsson replied:
Okay, sure, that makes sense!

jyn514 commented:
> I think ideally we would want two things for this, especially the first: Stabilizing HashMap::get_many_mut so we can be sure to match that API as much as possible.

Can IndexMap add it under a nightly or unstable opt-in cargo feature? Then people will have to opt in to breaking changes, and you can document that changes to this API won't result in a new major version.

(Or if you don't feel strongly, you can just make a new major release if the HashMap API changes for some reason.)

cuviper commented:
I don't like using nightly/unstable features, in part because it's an extra maintenance burden, but also because the feature may be opted-in on your behalf, arbitrarily deep in your dependency tree.

I'd rather not set us up for needing a semver bump either. If this is something that folks really need, let's apply that pressure on `std` stabilization. :)

cuviper commented:
Rust 1.86 (now in beta) will stabilize this as `get_disjoint_mut` with two different styles:

For `[T]`: https://doc.rust-lang.org/beta/std/primitive.slice.html#method.get_disjoint_mut
```rust
pub fn get_disjoint_mut<I, const N: usize>(
    &mut self,
    indices: [I; N],
) -> Result<[&mut <I as SliceIndex<[T]>>::Output; N], GetDisjointMutError>
where
    I: GetDisjointMutIndex + SliceIndex<[T]>,
```

For `HashMap<K, V>`: https://doc.rust-lang.org/beta/std/collections/struct.HashMap.html#method.get_disjoint_mut
```rust
pub fn get_disjoint_mut<Q, const N: usize>(
    &mut self,
    ks: [&Q; N],
) -> [Option<&mut V>; N]
where
    K: Borrow<Q>,
    Q: Hash + Eq + ?Sized,
```

We can decide for ourselves whether it makes sense to build on the standard library, or (unsafely) implement our own. `GetDisjointMutIndex` won't be stable though, nor the old `SliceIndex`, so if we want to replicate that support for both `usize` and ranges, then I think we'll be on our own anyway.

@NiklasJonsson it's been some time -- are you still interested in working on this?

cuviper commented:
> We can decide for ourselves whether it makes sense to build on the standard library, or (unsafely) implement our own. `GetDisjointMutIndex` won't be stable though, nor the old `SliceIndex`, so if we want to replicate that support for both `usize` and ranges, then I think we'll be on our own anyway.

Either way, I'd be fine with supporting just `usize` to start with, as long as there's a possibility to add ranges later. But even that could be separate, like how we have `get_index_mut` and `get_ranges_mut` covering the same as slice `get_mut` (and our `get_mut` matches `HashMap`). In other words, we could leave that extension to a future `get_disjoint_ranges_mut`.

NiklasJonsson commented:
Yeah I've been following the stabilization and I'd like to finish this up.
I'll have some time this weekend to have a deeper look but I think I'll
favor supporting less to start out and then I'll see if I have time to do
more.

On Tue, Feb 25, 2025, 02:29 Josh Stone ***@***.***> wrote:

> Rust 1.86 (now in beta) will stabilize this as get_disjoint_mut with two
> different styles:
>
> For [T]:
> https://doc.rust-lang.org/beta/std/primitive.slice.html#method.get_disjoint_mut
>
> pub fn get_disjoint_mut<I, const N: usize>(
>     &mut self,
>     indices: [I; N],) -> Result<[&mut <I as SliceIndex<[T]>>::Output; N], GetDisjointMutError>where
>     I: GetDisjointMutIndex + SliceIndex<[T]>,
>
> For HashMap<K, V>:
> https://doc.rust-lang.org/beta/std/collections/struct.HashMap.html#method.get_disjoint_mut
>
> pub fn get_disjoint_mut<Q, const N: usize>(
>     &mut self,
>     ks: [&Q; N],) -> [Option<&mut V>; N]where
>     K: Borrow<Q>,
>     Q: Hash + Eq + ?Sized,
>
> We can decide for ourselves whether it makes sense to build on the
> standard library, or (unsafely) implement our own. GetDisjointMutIndex
> won't be stable though, nor the old SliceIndex, so if we want to
> replicate that support for both usize and ranges, then I think we'll be
> on our own anyway.
>
> @NiklasJonsson <https://github.com/NiklasJonsson> it's been some time --
> are you still interested in working on this?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/indexmap-rs/indexmap/pull/238#issuecomment-2680128232>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADNRXP6VM3HKI4SGOGBRNOD2RPBONAVCNFSM6AAAAABXZNCFMWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBQGEZDQMRTGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: cuviper]*cuviper* left a comment (indexmap-rs/indexmap#238)
> <https://github.com/indexmap-rs/indexmap/pull/238#issuecomment-2680128232>
>
> Rust 1.86 (now in beta) will stabilize this as get_disjoint_mut with two
> different styles:
>
> For [T]:
> https://doc.rust-lang.org/beta/std/primitive.slice.html#method.get_disjoint_mut
>
> pub fn get_disjoint_mut<I, const N: usize>(
>     &mut self,
>     indices: [I; N],) -> Result<[&mut <I as SliceIndex<[T]>>::Output; N], GetDisjointMutError>where
>     I: GetDisjointMutIndex + SliceIndex<[T]>,
>
> For HashMap<K, V>:
> https://doc.rust-lang.org/beta/std/collections/struct.HashMap.html#method.get_disjoint_mut
>
> pub fn get_disjoint_mut<Q, const N: usize>(
>     &mut self,
>     ks: [&Q; N],) -> [Option<&mut V>; N]where
>     K: Borrow<Q>,
>     Q: Hash + Eq + ?Sized,
>
> We can decide for ourselves whether it makes sense to build on the
> standard library, or (unsafely) implement our own. GetDisjointMutIndex
> won't be stable though, nor the old SliceIndex, so if we want to
> replicate that support for both usize and ranges, then I think we'll be
> on our own anyway.
>
> @NiklasJonsson <https://github.com/NiklasJonsson> it's been some time --
> are you still interested in working on this?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/indexmap-rs/indexmap/pull/238#issuecomment-2680128232>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADNRXP6VM3HKI4SGOGBRNOD2RPBONAVCNFSM6AAAAABXZNCFMWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBQGEZDQMRTGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cuviper commented:
> 2\. We might _also_ want to wrap that into our own `get_many_index_mut`,

Correcting myself -- this should be plural, so now it would be `get_disjoint_indices_mut`.

NiklasJonsson commented:
@cuviper  Just to verify that I understand what you mean. An alright initial implementation would be:

1. The key-based `get_disjoint_mut`:
    ```rs
    pub fn get_disjoint_mut<Q, const N: usize>(
        &mut self,
        ks: [&Q; N],
    ) -> [Option<&mut V>; N]
    where
        K: Borrow<Q>,
        Q: Hash + Eq + ?Sized,
    ```
2. A `get_disjoint_indices_mut` that takes a usize array (pretty much the current `get_many_index_mut`A)

and then we can leave the implementation of `get_disjoint_mut` with ranges/for a future PR? If so, then it seems to be mostly about renaming the functions.

NiklasJonsson commented:
> Just to verify that I understand what you mean. An alright initial implementation would be:
> ...
I went ahead and implemented what I understood you to mean 🙂 Also added some extra tests.

cuviper commented on src/map.rs:
```diff
@@ -1196,6 +1227,44 @@ impl<K, V, S> IndexMap<K, V, S> {
+        let out = indices.map(|i| {
+            // SAFETY: The base pointer is valid as it comes from a slice and the deref is always
+            // in-bounds as we've already checked the indices above.
+            #[allow(unsafe_code)]
```
Nit: I think this `allow` should go around the entire method. While the actual `unsafe` operation is narrow, the SAFETY requirements need care in the whole thing, and the allowed lint is a good signal for that.

cuviper commented on src/map.rs:
````diff
@@ -790,6 +790,37 @@ where
+    /// let mut map = indexmap::IndexMap::from([(1, 'a'), (3, 'b'), (2, 'c')]);
+    /// assert_eq!(map.get_disjoint_mut([&2, &1]), Some([&mut 'c', &mut 'a']));
+    /// ```
+    pub fn get_disjoint_mut<Q, const N: usize>(&mut self, keys: [&Q; N]) -> Option<[&mut V; N]>
````
Also, `HashMap` panics if there's any overlap. The bounds-check panic you have is fine too, since that shouldn't happen normally, but possibly could if we have a logic error somewhere.

NiklasJonsson replied:
Aah right, I totally missed this part of the signature. I'll study the API in more detail.

cuviper replied:
Yeah, that's right, keep using the `Equivalent` bound.

NiklasJonsson replied:
>  so we can't directly rely on that, but we can probably have some internal method doing the unsafe part for either way.

The internal helper ended up like this:

```rs
pub(crate) fn get_disjoint_opt_mut<const N: usize>(
        &mut self,
        indices: [Option<usize>; N],
) -> Result<[Option<(&K, &mut V)>; N], GetDisjointMutError> {
```
and does most of the heavy-lifting for both of the methods. Does it seem alright to you? I don't feel great about wrapping indices in Some and then unwrapping after calling the helper in `Slice::get_disjoint_mut` but it does make the outer functions share most of the tricky code.

cuviper commented on src/map/slice.rs:
```diff
@@ -270,6 +271,52 @@ impl<K, V> Slice<K, V> {
+            };
+            if idx >= len {
+                return Err(GetDisjointMutError::IndexOutOfBounds);
+            } else if indices[i + 1..N].contains(&Some(idx)) {
```
It's valid either way, but it's a little simpler to compare with previous indices, and this is what the standard library does too in its `get_disjoint_check_valid`.
```suggestion
            } else if indices[..i].contains(&Some(idx)) {
```

cuviper commented on src/map.rs:
```diff
@@ -790,6 +790,33 @@ where
+    #[allow(unsafe_code)]
+    pub fn get_disjoint_mut<Q, const N: usize>(&mut self, keys: [&Q; N]) -> [Option<&mut V>; N]
+    where
+        Q: Hash + Equivalent<K> + ?Sized,
```
Just for consistency:
```suggestion
        Q: ?Sized + Hash + Equivalent<K>,
```

cuviper commented on src/lib.rs:
```diff
@@ -268,3 +268,33 @@ impl core::fmt::Display for TryReserveError {
+///
+/// It indicates one of two possible errors:
+/// - An index is out-of-bounds.
+/// - The same index appeared multiple times in the array
+///   (or different but overlapping indices when ranges are provided).
```
Let's hide the latter part, since we're not supporting that yet.
```suggestion
/// - The same index appeared multiple times in the array.
//    (or different but overlapping indices when ranges are provided)
```

cuviper commented on src/map.rs:
````diff
@@ -790,6 +790,33 @@ where
+    /// let mut map = indexmap::IndexMap::from([(1, 'a'), (3, 'b'), (2, 'c')]);
+    /// assert_eq!(map.get_disjoint_mut([&2, &1]), [Some(&mut 'c'), Some(&mut 'a')]);
+    /// ```
+    #[allow(unsafe_code)]
````
Is this outdated? Nothing in this function is unsafe.
```suggestion
```

cuviper commented on src/map/tests.rs:
```diff
@@ -828,3 +828,181 @@ move_index_oob!(test_move_index_out_of_bounds_0_10, 0, 10);
+    map.insert(321, 20);
+    assert_eq!(
+        map.get_disjoint_indices_mut([1, 2, 1]),
+        Err(crate::GetDisjointMutError::OverlappingIndices)
```
The `..i` change I suggested above makes this test fail, returning `IndexOutOfBounds` on the 2 instead. Either error would be justifiable here, but this test could change to `[1, 0, 1]` to avoid the OOB.

cuviper reviewed (approved):
Ok, Rust 1.86.0 shipped for us to match API, and I pushed the last few review suggestions myself. I think this is ready to go -- thank you for your contribution!

NiklasJonsson commented:
> Ok, Rust 1.86.0 shipped for us to match API, and I pushed the last few review suggestions myself. I think this is ready to go -- thank you for your contribution!

Thanks, happy to help!
